### PR TITLE
Feat 2857: Implement Gen2 Support for `ibm_database_task/tasks` datasource

### PR DESCRIPTION
**Pull request by pushpalathaeg**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Ticket : https://github.ibm.com/compose/App/issues/2857

# Gen2 Support for Database Task and Tasks Data Sources

## Overview
This documentation covers the enhancement to the IBM Cloud Terraform Provider that adds support for Generation 2 (Gen2) database instances in the `ibm_database_task` and `ibm_database_tasks` data sources. This update enables users to seamlessly query task information for both Classic and Gen2 database deployments using the same data sources with automatic detection and appropriate handling of generation-specific features.

### Gen2 Database Task Support in Data Sources
The `ibm_database_task` and `ibm_database_tasks` data sources now automatically detect whether a database instance is Classic or Gen2 and retrieve the appropriate task information. This provides a unified interface for querying database task status regardless of the database generation.

### Key Features
✅ **Automatic Detection**: The provider automatically identifies whether a database is Classic or Gen2 based on the plan type  
✅ **Unified Interface**: Use the same data sources for both Classic and Gen2 instances  
✅ **Backward Compatible**: Existing Terraform configurations continue to work without modification  
✅ **Proper State Management**: Gen2-specific behavior ensures clean state management when switching between instance types  
✅ **Comprehensive Testing**: Full test coverage for Gen2 scenarios including edge cases

### Classic Databases (Standard Plans)
- **Deployment Model**: Traditional IBM Cloud Databases architecture
- **Plan Names**: `standard`, `enterprise`, `platinum`
- **Task Management**: Full task history via Cloud Databases v5 API
- **Features**:
  - Detailed task information with unique task IDs
  - Complete task history
  - Task progress tracking
  - Multiple concurrent tasks

### Gen2 Databases (Standard-Gen2 Plans)
- **Deployment Model**: Next-generation architecture with enhanced capabilities
- **Plan Names**: `standard-gen2`, `enterprise-gen2`
- **Benefits**:
  - Enhanced performance and scalability
  - Improved resource isolation
  - Modern infrastructure
  - Better integration with IBM Cloud platform services
- **Task Management**: Derived from instance state via Resource Controller API
- **Approach**: Single task representing current instance state

### Backend Selection Pattern
The data sources use a strategy pattern to handle different database generations:

```
User Query → Read Function → Detect Plan Type → Select Backend
                                                      ↓
                                    ┌─────────────────┴─────────────────┐
                                    ↓                                   ↓
                          Classic Backend                        Gen2 Backend
                       (Cloud Databases v5 API)            (Resource Controller API)
```

### Implementation Components
- **data_source_ibm_database_task.go**: Main data source with backend selection logic
- **data_source_ibm_database_task_gen2.go**: Gen2 backend implementation for single task
- **data_source_ibm_database_tasks.go**: Main data source with backend selection logic
- **data_source_ibm_database_tasks_gen2.go**: Gen2 backend implementation for multiple tasks
- **Test Files**: Comprehensive test coverage for both generations

### Task ID Behavior

| Attribute | Classic Behavior | Gen2 Behavior | Reason |
|-----------|-----------------|---------------|---------|
| `task_id` | Returns unique task identifier | Returns empty string `""` | Gen2 doesn't expose task IDs via Resource Controller API |

### Task Representation

**Classic:**
- Returns actual task history
- Multiple tasks can be queried
- Each task has unique ID and detailed information

**Gen2:**
- Returns single "task" representing current instance state
- `tasks.#` is always `1` for Gen2
- Task information derived from instance metadata

### Status Mapping for Gen2

Gen2 instance states are mapped to task statuses:

| Instance State | Task Status | Description |
|---------------|-------------|-------------|
| `active` | `completed` | Instance is running normally |
| `provisioning`, `preparing` | `running` | Instance is being created or updated |
| `inactive` | `queued` | Instance is stopped or pending |
| `failed`, `removed` | `failed` | Instance encountered an error |

### Progress Calculation for Gen2

Progress percentage is calculated based on instance state:
- `active`: 100%
- `provisioning`, `preparing`: 50%
- `inactive`: 0%
- `failed`, `removed`: 0%

### Input Parameters - ibm_database_task

| Parameter | Type | Required | Description |
|-----------|------|----------|-------------|
| `task_id` | String | Yes | For Classic: Task ID; For Gen2: Database instance CRN |

### Input Parameters - ibm_database_tasks

| Parameter | Type | Required | Description |
|-----------|------|----------|-------------|
| `deployment_id` | String | Yes | Database instance CRN (works for both Classic and Gen2) |

### Output Attributes - Common (Both Generations)

| Attribute | Type | Description |
|-----------|------|-------------|
| `task_id` | String | Task identifier (empty for Gen2) |
| `deployment_id` | String | Database instance CRN |
| `status` | String | Task status (completed, running, queued, failed) |
| `description` | String | Task description |
| `progress_percent` | Integer | Task progress (0-100) |
| `created_at` | String | Task creation timestamp |

### Output Attributes - ibm_database_tasks Only

| Attribute | Type | Description |
|-----------|------|-------------|
| `tasks` | List | List of tasks (always 1 item for Gen2) |
| `tasks[].task_id` | String | Task identifier (empty for Gen2) |
| `tasks[].deployment_id` | String | Database instance CRN |
| `tasks[].status` | String | Task status |
| `tasks[].description` | String | Task description |
| `tasks[].progress_percent` | Integer | Task progress |
| `tasks[].created_at` | String | Task creation timestamp |

### Example 1: Query Single Task (Works for Both Generations)

```hcl
# Data source to query the task information for the Gen2 database
data "ibm_database_task" "database_task" {
  task_id = "ibm_database.my_database.id"
}

# Single output containing all task information
output "task_info" {
  description = "Complete task information for Gen2 database"
  value = {
    task_id          = data.ibm_database_task.database_task.task_id
    deployment_id    = data.ibm_database_task.database_task.deployment_id
    description      = data.ibm_database_task.database_task.description
    status           = data.ibm_database_task.database_task.status
    progress_percent = data.ibm_database_task.database_task.progress_percent
    created_at       = data.ibm_database_task.database_task.created_at
  }
}

```

### Example 2: Query Multiple Tasks (Works for Both Generations)

```hcl
# Query tasks for any database instance
data "ibm_database_tasks" "database_tasks" {
  deployment_id = "ibm_database.my_database.id"
}

# Access tasks information
output "tasks" {
  value = data.ibm_database_tasks.database_tasks.tasks
}

```

### Test Coverage
The implementation includes comprehensive test coverage for Gen2 scenarios:

#### Test Cases for ibm_database_task

**TestAccIBMDatabaseTaskGen2DataSourceBasic**
- Validates successful data retrieval for Gen2 instances
- Verifies common attributes (deployment_id, status, progress_percent)
- Confirms Gen2-specific behavior (empty task_id)
- Checks description and created_at fields

#### Test Cases for ibm_database_tasks

**TestAccIBMDatabaseTasksGen2DataSourceBasic**
- Validates successful data retrieval for Gen2 instances
- Verifies tasks list contains exactly 1 item
- Confirms Gen2-specific behavior (empty task_id in tasks)
- Checks all task attributes are properly set

</br>
</br>

**`ibm_database_task` TF  Tests for Gen2**

**TF INPUT CONFIG**
```
# Data source to query the task information for the Gen2 database
data "ibm_database_task" "database_task" {
  task_id = "crn:v1:bluemix:public:databases-for-postgresql:ca-mon:a/40ddc34a953a8c02f10987b59085b60e:6e752958-40f7-4f94-8799-0b7fd7120bb1::"
}


# Single output containing all task information
output "task_info" {
  description = "Complete task information for Gen2 database"
  value = {
    task_id          = data.ibm_database_task.database_task.task_id
    deployment_id    = data.ibm_database_task.database_task.deployment_id
    description      = data.ibm_database_task.database_task.description
    status           = data.ibm_database_task.database_task.status
    progress_percent = data.ibm_database_task.database_task.progress_percent
    created_at       = data.ibm_database_task.database_task.created_at
  }
}
```

**TF OUTPUT**
```
➜  ibm-database git:(2857/gen2-ibm_database_tasks-datasource) ✗ terraform apply

data.ibm_database_task.database_task: Reading...
data.ibm_database_task.database_task: Read complete after 4s [id=crn:v1:bluemix:public:databases-for-postgresql:ca-mon:a/40ddc34a953a8c02f10987b59085b60e:6e752958-40f7-4f94-8799-0b7fd7120bb1::]

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

task_info = {
  "created_at" = "2026-05-19T11:12:32.972Z"
  "deployment_id" = "crn:v1:bluemix:public:databases-for-postgresql:ca-mon:a/40ddc34a953a8c02f10987b59085b60e:6e752958-40f7-4f94-8799-0b7fd7120bb1::"
  "description" = "Provision completed successfully"
  "progress_percent" = 100
  "status" = "completed"
  "task_id" = ""
}
```

</br>
</br>

**`ibm_database_task` TF Tests for classic**

**TF INPUT CONFIG**
```
# Data source to query the task information for the Classic database
data "ibm_database_task" "database_task" {
  task_id = "crn:v1:bluemix:public:databases-for-postgresql:us-east:a/40ddc34a953a8c02f10987b59085b60e:3f5a3820-68d1-4f14-9936-581e1b3814f2:task:646de1c0-c444-4049-983f-d75fdc218d39"
}


# Single output containing all task information
output "task_info" {
  description = "Complete task information for Classic database"
  value = {
    task_id          = data.ibm_database_task.database_task.task_id
    deployment_id    = data.ibm_database_task.database_task.deployment_id
    description      = data.ibm_database_task.database_task.description
    status           = data.ibm_database_task.database_task.status
    progress_percent = data.ibm_database_task.database_task.progress_percent
    created_at       = data.ibm_database_task.database_task.created_at
  }
}
```

**TF OUTPUT**

```
ibm-database git:(2857/gen2-ibm_database_tasks-datasource) ✗ terraform apply
data.ibm_database_task.database_task: Reading...
data.ibm_database_task.database_task: Read complete after 6s [id=crn:v1:bluemix:public:databases-for-postgresql:us-east:a/40ddc34a953a8c02f10987b59085b60e:3f5a3820-68d1-4f14-9936-581e1b3814f2:task:646de1c0-c444-4049-983f-d75fdc218d39]

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

task_info = {
  "created_at" = "2026-06-11T06:17:01.000Z"
  "deployment_id" = "crn:v1:bluemix:public:databases-for-postgresql:us-east:a/40ddc34a953a8c02f10987b59085b60e:3f5a3820-68d1-4f14-9936-581e1b3814f2::"
  "description" = "Creating  backup"
  "progress_percent" = 33
  "status" = "failed"
  "task_id" = "crn:v1:bluemix:public:databases-for-postgresql:us-east:a/40ddc34a953a8c02f10987b59085b60e:3f5a3820-68d1-4f14-9936-581e1b3814f2:task:646de1c0-c444-4049-983f-d75fdc218d39"
}
```

</br>
</br>

**`ibm_database_tasks` TF  Tests for Classic**

**TF INPUT CONFIG**
```
# Data source to query the task information for the Classic database
data "ibm_database_tasks" "database_tasks" {
  deployment_id = "crn:v1:bluemix:public:databases-for-postgresql:us-east:a/40ddc34a953a8c02f10987b59085b60e:3f5a3820-68d1-4f14-9936-581e1b3814f2::"
}


output "tasks" {
  value = data.ibm_database_tasks.database_tasks.tasks
}
```

**TF OUTPUT**
```
ibm-database git:(2857/gen2-ibm_database_tasks-datasource) ✗ terraform apply
Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

tasks = tolist([
  {
    "created_at" = "2026-06-15T06:02:29.000Z"
    "deployment_id" = "crn:v1:bluemix:public:databases-for-postgresql:us-east:a/40ddc34a953a8c02f10987b59085b60e:3f5a3820-68d1-4f14-9936-581e1b3814f2::"
    "description" = "Creating an on-demand backup"
    "progress_percent" = 100
    "status" = "completed"
    "task_id" = "crn:v1:bluemix:public:databases-for-postgresql:us-east:a/40ddc34a953a8c02f10987b59085b60e:3f5a3820-68d1-4f14-9936-581e1b3814f2:task:14a52602-9b58-45cc-b0cf-2b5bb420cf52"
  },
  {
    "created_at" = "2026-06-14T06:17:02.000Z"
    "deployment_id" = "crn:v1:bluemix:public:databases-for-postgresql:us-east:a/40ddc34a953a8c02f10987b59085b60e:3f5a3820-68d1-4f14-9936-581e1b3814f2::"
    "description" = "Creating  backup"
    "progress_percent" = 100
    "status" = "completed"
    "task_id" = "crn:v1:bluemix:public:databases-for-postgresql:us-east:a/40ddc34a953a8c02f10987b59085b60e:3f5a3820-68d1-4f14-9936-581e1b3814f2:task:9d2fb875-723e-4e19-b9e7-faf2fb5c029e"
  },
])
```

</br>
</br>

</br>
</br>

**`ibm_database_tasks` TF  Tests for Gen2**

**TF INPUT CONFIG**
```
# Data source to query the task information for the Gen2 database
data "ibm_database_tasks" "database_tasks" {
  deployment_id = "crn:v1:bluemix:public:databases-for-postgresql:ca-mon:a/40ddc34a953a8c02f10987b59085b60e:7b231067-9ddd-4de9-964a-d7bf84cfdc3f::"
}


output "tasks" {
  value = data.ibm_database_tasks.database_tasks.tasks
}

```

**TF OUTPUT**
```
 ibm-database git:(2857/gen2-ibm_database_tasks-datasource) ✗ terraform apply
Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

tasks = tolist([
  {
    "created_at" = "2026-06-02T12:28:44.963Z"
    "deployment_id" = "crn:v1:bluemix:public:databases-for-postgresql:ca-mon:a/40ddc34a953a8c02f10987b59085b60e:7b231067-9ddd-4de9-964a-d7bf84cfdc3f::"
    "description" = "Instance update is completed."
    "progress_percent" = 100
    "status" = "completed"
    "task_id" = ""
  },
])
```

**DOC update**

Screenshot 2026-09-21 at 11 54 25 AM

Screenshot 2026-09-21 at 11 57 34 AM

Screenshot 2026-09-21 at 11 54 41 AM

Screenshot 2026-09-21 at 11 59 11 AM

Output from acceptance testing:

```

export ICD_DB_GEN2_DEPLOYMENT_ID
export ICD_DB_GEN2_TASK_ID
export TF_ACC=true

$ go test -v -run "TestAccIBMDatabaseTaskGen2DataSourceBasic" -timeout 10m
=== RUN   TestAccIBMDatabaseTaskGen2DataSourceBasic
--- PASS: TestAccIBMDatabaseTaskGen2DataSourceBasic (31.87s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/database        34.710s


go test -v -run "TestAccIBMDatabaseTasksGen2DataSourceBasic" -timeout 10m
=== RUN   TestAccIBMDatabaseTasksGen2DataSourceBasic
--- PASS: TestAccIBMDatabaseTasksGen2DataSourceBasic (29.61s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/database        32.403s
...
```

### CODE REVIEW

File: ibm/service/database/data_source_ibm_database_task_gen2.go [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+	if instance.State == nil {
+		return 0
+	}
+
+	state := *instance.State
+	switch state {
+	case "active":
+		return 100
+	case "provisioning":
+		return 50
```
Comment (uibm) [inline ibm/service/database/data_source_ibm_database_task_gen2.go]: are these numbers random ?

File: ibm/service/database/data_source_ibm_database_tasks_gen2.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+		return flex.DateTimeToString(instance.UpdatedAt)
+	}
+	if instance.CreatedAt != nil {
+		return flex.DateTimeToString(instance.CreatedAt)
+	}
+
+	return time.Now().UTC().Format(time.RFC3339)
+}
+
+// Made with Bob
```
Comment (uibm) [inline ibm/service/database/data_source_ibm_database_tasks_gen2.go]: This we can get rid of

File: ibm/service/database/data_source_ibm_database_task_gen2.go (hunk 1 above)
Comment (pushpalathaeg) [reply ibm/service/database/data_source_ibm_database_task_gen2.go]: @uibm : calculateProgress method will return the value based on instance status for the Gen2 task

File: ibm/service/database/data_source_ibm_database_tasks_gen2.go (hunk 2 above)
Comment (pushpalathaeg) [reply ibm/service/database/data_source_ibm_database_tasks_gen2.go]: addressed

File: ibm/service/database/data_source_ibm_database_task_gen2.go (hunk 1 above)
Comment (obai-1) [reply ibm/service/database/data_source_ibm_database_task_gen2.go]: Is this consistent with what web-ui is showing today?

File: ibm/service/database/data_source_ibm_database_task_gen2.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+	state := *instance.State
+	switch state {
+	case "active":
+		return "completed"
+	case "provisioning", "in progress":
+		return "running"
+	case "failed":
+		return "failed"
+	case "inactive":
+		return "queued"
```
Comment (obai-1) [inline ibm/service/database/data_source_ibm_database_task_gen2.go]: Is this correct?

Comment (pushpalathaeg) [reply ibm/service/database/data_source_ibm_database_task_gen2.go]: "inactive" → "inactive" - scenarios for this Instance is stopped or suspended, so keep original state

File: ibm/service/database/data_source_ibm_database_task_gen2.go (hunk 1 above)
Comment (pushpalathaeg) [reply ibm/service/database/data_source_ibm_database_task_gen2.go]: @obai-1  : In web-ui :
Displays task description, status icon, and time elapsed (using TimeAgo component)
No percentage or progress bar is used

For TF I followed like below
-  nil state → 0% - No state information available
- "active" → 100% - Instance is fully provisioned and operational - 100% complete
- "provisioning" → 50% - Instance is being created - estimated at 50% (midpoint of provisioning process). Note: Actual progress may vary; RC API doesn't provide granular progress data
- "in progress" → 75% - Operation is in progress - estimated at 75% (nearing completion). Note: This is an approximation as RC API doesn't provide actual progress percentage
- "failed", "removed" → 100% - Operation has completed (either failed or instance removed) - 100% done
- "inactive" → 0% - Instance is stopped/suspended - no progress (0%)
- default → 0% - Unknown state - assume no progress

File: ibm/service/database/data_source_ibm_database_task_gen2.go (hunk 3 above)
Comment (pushpalathaeg) [reply ibm/service/database/data_source_ibm_database_task_gen2.go]: @obai-1 : addressed it

File: ibm/service/database/data_source_ibm_database_task.go [hunk 4]
```diff
@@ -6,15 +6,82 @@ package database
...
+		}
+
+		// Get the instance using the deployment_id from the task
+		instance, _, err = rsConClient.GetResourceInstance(&rc.GetResourceInstanceOptions{
+			ID: &deploymentID,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("failed to get resource instance: %s", err)
+		}
+	}
```
Comment (obai-1) [inline ibm/service/database/data_source_ibm_database_task.go]: You can introspect the task_id and get the deployment_id directly without doing this trial and error.

File: ibm/service/database/data_source_ibm_database_task_gen2.go [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+	}
+
+	if !isGen2Plan(*instance.ResourcePlanID) {
+		tfErr := flex.TerraformErrorf(
+			fmt.Errorf("instance is not a Gen2 database"),
+			"this data source is for Gen2 databases only",
+			"(Data) ibm_database_task",
+			"read")
+		return tfErr.GetDiag()
+	}
```
Comment (obai-1) [inline ibm/service/database/data_source_ibm_database_task_gen2.go]: Any reason to do this check again here? Before this Gen2 Read function is called, we are checking if it is Gen2, right?

File: ibm/service/database/data_source_ibm_database_task_gen2.go [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+	log.Printf("[DEBUG] RC Instance Details for Gen2 Task:")
+	if instance.ID != nil {
+		log.Printf("[DEBUG]   ID: %s", *instance.ID)
+	}
+	if instance.State != nil {
+		log.Printf("[DEBUG]   State: %s", *instance.State)
+	}
+	if instance.LastOperation != nil {
+		log.Printf("[DEBUG]   LastOperation: %+v", instance.LastOperation)
+	}
```
Comment (obai-1) [inline ibm/service/database/data_source_ibm_database_task_gen2.go]: Any reason to leave these logs here? Instead of printing DEBUG messages like this, move the logs to the area where you are using them.

File: ibm/service/database/data_source_ibm_database_task_gen2.go [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+		return "completed"
+	case "provisioning", "in progress":
+		// Instance is being created or an operation is in progress
+		return "running"
+	case "failed":
+		// Instance provisioning or operation has failed
+		return "failed"
+	case "inactive":
+		// Instance is stopped or suspended, keep original state
+		return "inactive"
```
Comment (obai-1) [inline ibm/service/database/data_source_ibm_database_task_gen2.go]: If you remove this the `default` case will handle them, right?

Comment (pushpalathaeg) [reply ibm/service/database/data_source_ibm_database_task_gen2.go]: @obai-1 : Yes, addressed it

File: ibm/service/database/data_source_ibm_database_task_gen2.go (hunk 6 above)
Comment (pushpalathaeg) [reply ibm/service/database/data_source_ibm_database_task_gen2.go]: @obai-1  Yes moved

File: ibm/service/database/data_source_ibm_database_task_gen2.go (hunk 5 above)
Comment (pushpalathaeg) [reply ibm/service/database/data_source_ibm_database_task_gen2.go]: @obai-1  : Yes, The Gen2 check (isGen2Plan) is already performed in data_source_ibm_database_task.go before instantiating the Gen2 backend so Removed redundant Gen2 validation code

File: ibm/service/database/data_source_ibm_database_task.go (hunk 4 above)
Comment (pushpalathaeg) [reply ibm/service/database/data_source_ibm_database_task.go]: @obai-1 :  Simplified Backend Selection with CRN Introspection.

- Introspects the task_id CRN format to determine deployment ID:
             Classic CRN (contains `:task:`): Extract deployment_id from task 
             Gen2 CRN (no `:task:` segment): Use task_id directly as deployment_id
- Eliminated trial-and-error approach (no more failed API calls and retries)
- Reduced code duplication by using a single GetResourceInstance call

File: ibm/service/database/data_source_ibm_database_tasks_gen2.go [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+	}
+
+	if !isGen2Plan(*instance.ResourcePlanID) {
+		tfErr := flex.TerraformErrorf(
+			fmt.Errorf("instance is not a Gen2 database"),
+			"this data source is for Gen2 databases only",
+			"(Data) ibm_database_tasks",
+			"read")
+		return tfErr.GetDiag()
+	}
```
Comment (obai-1) [inline ibm/service/database/data_source_ibm_database_tasks_gen2.go]: We are already in Gen2 read. Do we need to check this again?

File: ibm/service/database/data_source_ibm_database_tasks_gen2.go [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+		// Instance is being created or an operation is in progress
+		return "running"
+	case "removed":
+		// Instance has been deleted, operation is complete
+		return "completed"
+	default:
+		// Return the original state for any unmapped states
+		return state
+	}
+}
```
Comment (obai-1) [inline ibm/service/database/data_source_ibm_database_tasks_gen2.go]: Looks like this code/logic is in another place as well. Can you unify and avoid duplication ?

File: ibm/service/database/data_source_ibm_database_tasks_gen2.go (hunk 8 above)
Comment (pushpalathaeg) [reply ibm/service/database/data_source_ibm_database_tasks_gen2.go]: @obai-1 Yes Its not need to add it again so removed the duplication

File: ibm/service/database/data_source_ibm_database_tasks_gen2.go (hunk 9 above)
Comment (pushpalathaeg) [reply ibm/service/database/data_source_ibm_database_tasks_gen2.go]: @obai-1 Yes removed the duplication and unified the logic

File: ibm/service/database/gen2_task_utils.go [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+	"fmt"
+	"time"
+
+	"github.com/IBM-Cloud/terraform-provider-ibm/ibm/flex"
+	rc "github.com/IBM/platform-services-go-sdk/resourcecontrollerv2"
+)
+
+// gen2TaskUtils provides shared utility functions for Gen2 database task operations.
+// These functions are used by both ibm_database_task and ibm_database_tasks data sources.
+type gen2TaskUtils struct{}
```
Comment (obai-1) [inline ibm/service/database/gen2_task_utils.go]: This type seems unnecessary. It's just adding complexity without any benefit. Do you see any reason to have it?

File: ibm/service/database/data_source_ibm_database_tasks_gen2.go [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+	"github.com/IBM-Cloud/terraform-provider-ibm/ibm/conns"
+	"github.com/IBM-Cloud/terraform-provider-ibm/ibm/flex"
+	rc "github.com/IBM/platform-services-go-sdk/resourcecontrollerv2"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+// dataSourceIBMDatabaseTasksGen2Backend implements tasks retrieval for Gen2 databases using RC API.
+type dataSourceIBMDatabaseTasksGen2Backend struct {
+	utils gen2TaskUtils
```
Comment (obai-1) [inline ibm/service/database/data_source_ibm_database_tasks_gen2.go]: If you remove the type, as mentioned here https://github.com/IBM-Cloud/terraform-provider-ibm/pull/6821/changes#r3670410117, then you can remove this line

File: ibm/service/database/gen2_task_utils.go (hunk 10 above)
Comment (pushpalathaeg) [reply ibm/service/database/gen2_task_utils.go]: @obai-1   Here's a summary of every change made across the three files:
[gen2_task_utils.go](vscode-webview://0g4hhki6omjbj27u95v4ru86d44brkih07cuu9353no0mlki9p70/ibm/service/database/gen2_task_utils.go) — removed type gen2TaskUtils struct{} and converted all four methods to plain package-level functions with a gen2 prefix:
* (g *gen2TaskUtils) getOperationDescription → gen2GetOperationDescription
* (g *gen2TaskUtils) mapStateToStatus → gen2MapStateToStatus
* (g *gen2TaskUtils) calculateProgress → gen2CalculateProgress
* (g *gen2TaskUtils) getOperationTime → gen2GetOperationTime
[data_source_ibm_database_tasks_gen2.go](vscode-webview://0g4hhki6omjbj27u95v4ru86d44brkih07cuu9353no0mlki9p70/ibm/service/database/data_source_ibm_database_tasks_gen2.go) — removed the utils gen2TaskUtils field from the struct; updated four g.utils.* call sites to direct function calls.
[data_source_ibm_database_task_gen2.go](vscode-webview://0g4hhki6omjbj27u95v4ru86d44brkih07cuu9353no0mlki9p70/ibm/service/database/data_source_ibm_database_task_gen2.go) — same: removed the utils field, updated four call sites.

File: ibm/service/database/data_source_ibm_database_tasks_gen2.go (hunk 11 above)
Comment (pushpalathaeg) [reply ibm/service/database/data_source_ibm_database_tasks_gen2.go]: @obai-1 Yes, removed here also

File: ibm/service/database/gen2_task_utils.go [hunk 12]
```diff
@@ -0,0 +1,101 @@
+// Copyright IBM Corp. 2024 All Rights Reserved.
```
Comment (obai-1) [inline ibm/service/database/gen2_task_utils.go]: This file name is inconsistent gen2_task_utils.go
See if you can combine this with helpers.go

Comment (pushpalathaeg) [reply ibm/service/database/gen2_task_utils.go]: @obai-1  removed the gen2_task_utils.go file and moved the methods to helpers.go

**Review by obai-1 — APPROVED** [review]
